From 009da21cfda69c98c90efd93ad6c70621eb0935c Mon Sep 17 00:00:00 2001
From: Jeevaka Prabu Badrappan <jeevaka.badrappan@intel.com>
Date: Fri, 8 Apr 2022 20:35:04 +0530
Subject: [PATCH] Address error logs during youtube video playback

Eventhough, youtube video playback works using hardware decoder
error log "SyncOperation failed" can be seen on each frame header.

When DecodeFrameAsync fails with MFX_WRN_DEVICE_BUSY, as per
documentation SyncOperation can be called or DecodeFrameAsync can
be called after few milliseconds. Since the syncpoint is always
NULL when DecodeFrameAsync fails with DEVICE_BUSY logging it as
error log is incorrect. Also, instead of calling DecodeFrameAsync
immediately let it be called on next iteration.

Change-Id: I85756107bfa5658230ebb42846dc0651fba31aa4
Tracked-On: OAM-101884
Signed-off-by: Jeevaka Prabu Badrappan <jeevaka.badrappan@intel.com>
Signed-off-by: Ratnesh Kumar Rai <ratnesh.kumar.rai@intel.com>
---
 omx_components/src/mfx_omx_vdec_component.cpp | 25 ++++++++++++++++---
 1 file changed, 21 insertions(+), 4 deletions(-)

diff --git a/omx_components/src/mfx_omx_vdec_component.cpp b/omx_components/src/mfx_omx_vdec_component.cpp
index 4df0897..f3e02ec 100644
--- a/omx_components/src/mfx_omx_vdec_component.cpp
+++ b/omx_components/src/mfx_omx_vdec_component.cpp
@@ -3226,14 +3226,31 @@ mfxStatus MfxOmxVdecComponent::DecodeFrame(void)
             {
                 mfxStatus mfx_sts = MFX_ERR_NONE;
                 MFX_OMX_AUTO_TRACE("mfx(DecodeFrameAsync)::MFX_WRN_DEVICE_BUSY");
+                mfxSyncPoint* pDevBusySyncPoint = m_pSurfaces->GetSyncPoint();
+                if (pDevBusySyncPoint)
+                {
 #ifdef USE_ONEVPL
-                mfx_sts = MFXVideoCORE_SyncOperation(m_Session, *pSyncPoint, MFX_OMX_INFINITE);
+                    mfx_sts = MFXVideoCORE_SyncOperation(m_Session, *pDevBusySyncPoint, MFX_OMX_INFINITE);
 #else
-                mfx_sts = m_Session.SyncOperation(*pSyncPoint, MFX_OMX_INFINITE);
+                    mfx_sts = m_Session.SyncOperation(*pDevBusySyncPoint, MFX_OMX_INFINITE);
 #endif
+                }
+                else
+                {
+                    mfx_sts = MFX_ERR_NULL_PTR;
+                    m_pDevBusyEvent->TimedWait(1);
+                    m_pDevBusyEvent->Reset();
+                }
                 if (MFX_ERR_NONE != mfx_sts)
                 {
-                    MFX_OMX_LOG_ERROR("SyncOperation failed - %s", mfx_omx_code_to_string(mfx_sts));
+                    /*
+                     * Log commented out as MFX_ERR_NULL_PTR is not considered as exit failure in
+                     * DecodeFrame loop. This cerror happens on each frame header. Enabling this
+                     * log results in too many logs even when the subsequent call to
+                     * DecodeFrameAsync succeeds without any error.
+                     *
+                     * MFX_OMX_LOG_ERROR("SyncOperation failed - %s", mfx_omx_code_to_string(mfx_sts));
+                     */
                     mfx_res = mfx_sts;
                 }
             }
@@ -3318,7 +3335,7 @@ mfxStatus MfxOmxVdecComponent::DecodeFrame(void)
             m_pFreeSyncPoint = NULL;
             break;
         }
-        else
+        else if (MFX_ERR_NULL_PTR != mfx_res)
         {
             MFX_OMX_LOG_ERROR("DecodeFrameAsync failed - %s", mfx_omx_code_to_string(mfx_res));
         }
-- 
2.17.1

